DM-24829: Use a magic string for no component in datastore record

timj opened the pull request:
We were using an empty string but Oracle forces those
to be NULL and that breaks things.

timj committed:
Use a magic string for no component in datastore record

We were using an empty string but Oracle forces those
to be NULL and that breaks things.

MichelleGower reviewed (approved):
I did not search for database admin recommendations to see if they differ from this solution.  I started running ci_hsc_gen3 with Oracle using this ticket branch.   It has made it past running ISR at the time of this comment.  Dug around the output repo filenames a little bit to make sure this string didn't end up in a filename and couldn't find any.  Approving the changes for merging.

timj committed:
Add new constraint for 0 length strings to sqlite

Oracle converts zero-length strings to NULL and this can cause
unexpected problems there since sqlite does not do that. Modify
the sqlite constraints to check for this so we can spot the
problems before using Oracle.

timj committed:
Use None for default dataset type name to match Oracle

The name can be null and Oracle forces this, so to be consistent
use NULL for all backends.

timj commented on python/lsst/daf/butler/datastores/fileLikeDatastore.py:
```diff
@@ -62,6 +62,9 @@
 log = logging.getLogger(__name__)
 
+# String to use when a Python None is encountered
+NULLSTR = "__NULL_STRING__"
```
@TallJimbo I am pondering whether this should be shorter. Will people want to use this constant elsewhere at some point?

TallJimbo replied:
Let's not worry about it.  I think if we did need to use sentinal empty strings in more places, we'd move more than just the constant into `Database`, and I don't have any anticipated use cases of that right now.

timj committed:
Remove f-strings for strings that have no placeholders

TallJimbo reviewed (approved):
Looks good.

Thank goodness the f-string lint didn't affect multi-line strings where some of the lines have substitutions and others don't (where I vastly prefer to start all lines with `f`).
